winVersion/Windows 10 version fetch: use Windows Registry instead of a static release names to builds

josephsl opened the pull request:
Hi,

Mostly in preparation for a possibility that Windows 10 Version 21H2 will be used by multiple builds, including Server 2022, and prepares NVDA to support Windows 11 and future Windows versions:

### Link to issue number:
Closes #12509 

### Summary of the issue:
Currently a static Windows 10 release names to builds map is used to record Windows 10 builds to feature updates (releases). Until 21H1 there was a one-to-one correspondence between releases and builds, but it appears Version 21H2 will be represented by multiple builds:

* Windows 10: 19044
* Server 2022: 20348
* Windows 11: build 22000

Builds subject to confirmation.

Therefore this breaks current static map.

### Description of how this pull request fixes the issue:
Since Version 1511, Windows Registry holds release name which corresponds to builds (e.g. 1511 = 10586, 20H2 = 19042), housed inside a "Release Id" key. Starting with Version 20H2, a separate "DisplayVersion" key is used to announce public release name. In Windows Insider Preview builds, dev channel has display version of "Dev", making it even easier to determine what kind of build a user is running. This tradition continues with Windows 11 and Server 2022, with Registry recording "21H2", which makes sense given their projected release period. Therefore use Windows Registry instead of the static releases to builds map and deprecate releases to builds map.

### Testing strategy:
Manual testing:

1. On supported Windows 10 releases (1507 LTS, 1607/server 2016 LTS, 1809/Server 2019 LTS, 1909 (Enterprise and Education), 2004, 20H2, 21H1, 21H2/Server 2022 preview, Windows Insider Preview dev channel), open Windows Registry path described in #12509 and confirm that release Id's and display name keys are present (the latter is available starting with 20H2).
2. In Python Console, verify that winVersion.getWinVer() says "Windows 10 YYMM/YYHn" and the last part (release name) matches the value obtained from Windows Registry.
3. Optionally, define a local unit test that checks the version string returned by winVersion.getWinVer() starts with "Windows 10 release" for the local system (e.g. "Windows 10 21H1" on build 19043).

### Known issues with pull request:
None

### Change log entries:
Likely a change for developers, stating that on Windows 10, release names are obtained from Windows Registry, although note that this is an internal operation. A visible changelog is:

Changes for developers:

On Windows 10 Version 1511 and later (including Insider Preview builds), current Windows feature update release name is obtained from Windows Registry. (#12509)
The releases to builds mapping found in winVersion module is deprecated and will be removed in 2022.1. (#12544)

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual testing.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers

### Discussion:
If this PR is adopted, what should we do about the static release names to builds map? There are three possibilities:

1. Swap releases with builds so the map records build : release. This makes Windows Registry method unnecessary - adding future Windows releases will continue to involve editing this map.
2. Deprecate and remove the static map - the only place the map is used is when fetching Windows 10 release names anyway, and since Windows Registry contains needed information, there is no need for this map. If deprecation and removal are chosen, a follow-up PR should be created during 2022.1 development cycle to remove the static map altogether.
3. For releases represented by multiple builds, have Windows 10 release name fetcher check for these builds and return the release name for these. This makes version fetch routine complicated as one needs to hunt for all build numbers represented by a single release name.

Update (2021-07-07): deprecated and replaced by a private builds to releases mapping.

Thanks.

josephsl committed:
winVersion: use Windows Registry to fetch Windows 10 releaess. Re #12509.

From Widows 10 Version 1511 (November Update/build 10586), WinVer (About Windows) dialog uses Windows Registry to obtain version information (build 10240 wsa retroactively named '1507' then). Initially 'ReleaseId' key was used, but since 20H2, 'DisplayVersion' is used (ReleaseId for 20H2 is 2009).
The Registry method is needed to support multiple Windows 10 builds with same display version. Currently release Id's to builds map contains one to one mapping between versions and builds, whereas in 21H2 development cycle, up to three builds may represent this release. Therefore use Registry method to detect all public Windows 10 releases.

josephsl committed:
winVersion: releaseID -> releaseId, clarify comment. Re #12509

josephsl commented:
Hi,

By the way, people watching NVDA add-ons may notice a similarity with Resource Monitor - in fact, this PR is a simplified version of Windows 10 version fetch routine found in that add-on, therefore it went through years of real-life usage.

Thanks.

cary-rowen commented:
@josephsl 
There is a digression, Windows 10 application enhancement is so versatile; after years of extensive use and testing, why not integrate it into NVDA itself?
Thanks.

josephsl commented:
Hi, not all features from that add-on can be integrated into NVDA because Windows 10 apps (built-in and from Microsoft Store) can change without notice, which also applies to features found in Windows Insider Preview builds (things come and go without notice, making it harder to say something is stable enough to be part of NVDA unless it’s been out there for a while). To be honest, Windows 10 App Essentials is “versatile” because of the state of app and feature accessibility and usability across Windows 10 releases (many commits in this add-on are frustrations about app accessibility and usability issues, which have lessened somewhat in recent years). Another key consideration is support duration for Windows 10 releases – some features are part of Windows 10 releases no longer supported by Microsoft, therefore if support for them ends in the add-on, I promptly remove the code associated with it (the add-on is currently going through such a transition to fully remove support for releases earlier than Version 2004). Thanks.

cary-rowen commented:
Thanks @josephsl for reply

leonardder commented on source/winVersion.py:
```diff
@@ -49,12 +50,22 @@ def _windowsVersionToReleaseName(self):
+			# Always return "Windows 10 1507" on build 10240.
+			if self.build == 10240:
+				return "Windows 10 1507"
+			currentVersion = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows NT\CurrentVersion")
```
Please use a with clause

LeonarddeR reviewed (changes requested):
I assume it will suffice to remove the release map altogether.

josephsl commented:
Hi, I once saw Registry access used as a context manager but wasn’t sure if it is applicable; now you mention it, I’ll perform some tests and make necessary changes (this is also applicable to Resource Monitor add-on, come to think of it). Thanks for this tip.

josephsl committed:
winVersion/Windows 10 version fetch: use Windows Registry context manager. re #12509.

Advised by Leonard de Ruijter: use context manager (with) associated with Windows Registry HKEY handles (of the form 'with winreg.openKey(HKEY, path) as something').

josephsl commented:
Hi,

In the end, there is no point in keeping something that will soon break (look at various 21H2 build branches), so the releases map will be gone.

Thanks.

josephsl committed:
winVersion/Windows 10 releases: remove unused releases to builds map. Re #12509.

Windows 10 releases to builds map records feature update release names to builds (e.g. '1903': 18362). With the Windows Registry method in place, coupled with several build branches for Version 21H2, it makes sense to remove an unused dictionary.

feerrenrut commented on source/winVersion.py:
```diff
@@ -49,12 +50,23 @@ def _windowsVersionToReleaseName(self):
-			else:
-				# Windows Insider build.
-				return "Windows 10 prerelease"
+			# From Version 1511 (build 10586), release Id/display version comes from Windows Registry.
+			# Always return "Windows 10 1507" on build 10240.
+			if self.build == 10240:
+				return "Windows 10 1507"
+			with winreg.OpenKey(
+				winreg.HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows NT\CurrentVersion"
+			) as currentVersion:
+				# Version 20H2 and later where a separate display version string is used.
+				# For backward compatibility, release Id will store display version string.
+				try:
+					releaseId = winreg.QueryValueEx(currentVersion, "DisplayVersion")[0]
+				except OSError:
+					releaseId = None
+				# Version 1511 and later unless display version string is present.
+				if not releaseId:
+					releaseId = winreg.QueryValueEx(currentVersion, "ReleaseID")[0]
+			return f"Windows 10 {releaseId}"
```
This will return the name of the Windows release you are running on regardless of how the `WinVersion` class is constructed right?

Eg, ensure there is a unit test for this. I suspect this will fail on any computer running a later version of Windows.
```python
assertEqual(repr(WIN10_1511), "Windows 10 10586")
```

feerrenrut replied:
If you keep this for the `getWinVer` case, please cache it at module level, it doesn't need to happen more than once.

feerrenrut commented on source/winVersion.py:
```diff
@@ -124,21 +136,5 @@ def isUwpOcrAvailable():
 	return os.path.isdir(UWP_OCR_DATA_PATH)
 
 
-WIN10_RELEASE_NAME_TO_BUILDS = {
```
This is part of the API, it can be deprecated, but shouldn't be removed.

feerrenrut replied:
You'll need to restore this dictionary and add a comment that it is deprecated to be removed in 2022.1. You can also add a comment that it isn't used internally and should not be updated further.

feerrenrut replied:
Could you make this happen automatically, like this:
``` python
# Deprecated: Windows 10 releases will be obtained from Windows Registry, no entries will be added.
# WIN10_RELEASE_NAME_TO_BUILDS will be removed in 2022.1.
if version_year < 2022:
	WIN10_RELEASE_NAME_TO_BUILDS = {
```

feerrenrut commented on source/winVersion.py:
```diff
@@ -49,12 +50,23 @@ def _windowsVersionToReleaseName(self):
 		elif (self.major, self.minor) == (6, 3):
 			return "Windows 8.1"
 		elif self.major == 10:
-			buildsToReleases = {build: release for release, build in WIN10_RELEASE_NAME_TO_BUILDS.items()}
```
If this is the only usage, and it reverses the mapping, you could consider adding a new mapping, deprecating the old one. The new mapping would be internal only, and map builds to release names. Though this isn't necessary if we just construct the `class WinVersion` with the release name populated.

feerrenrut reviewed (changes requested):
I'm ok with getting the Release name like this (from the registry) when the `getWinVer` method is called. But for the constants it would be better to stick to an in code mapping, perhaps instead of the `WIN10_RELEASE_NAME_TO_BUILDS` dictionary, add an argument to `class WinVersion` to set the release name and set all the names with the constants. Then there can be a constant for each variation of  "21H2 " EG something like `WIN10_21H2_SERVER_2022 = WinVersion(major=10, minor=0, build=20348, releaseName="Windows 10 21H2 Server 2022")`

josephsl commented:
> I'm ok with getting the Release name like this (from the registry) when the `getWinVer` method is called. But for the constants it would be better to stick to an in code mapping, perhaps instead of the `WIN10_RELEASE_NAME_TO_BUILDS` dictionary, add an argument to `class WinVersion` to set the release name and set all the names with the constants. Then there can be a constant for each variation of "21H2 " EG something like `WIN10_21H2_SERVER_2022 = WinVersion(major=10, minor=0, build=20348, releaseName="Windows 10 21H2 Server 2022")`

Hi,

Adding "releaseName": this will also benefit Windows 7 and 8.x.

Windows Registry and release name: yes, Windows Registry will return the correct release name depending on the version you've got, so it can be used in winVersion.getWinVer function, I was wondering about unit tests - I'll come up with soething.

Release names to builds map: not needed as Windows Registry and release name argument will suffice.

Caching release name from the module: right, that is a useful optimization strategy.

Would you like me to restore the release names to builds dictionary for 2021.x and mark it as deprecated in the process?

Thanks.

josephsl committed:
winVersion.WinVersion class: define 'releaseName' attribute. Re #12509.

Advice from Reef Turner (NV Access): define a 'release name' atribute to set the public facing name for a Windows release such as 'Windows 7' for 6.1.7600. This is useful for Widows 10 due to many major updates released since July 2015.

josephsl committed:
WinVersion: return release name if defined for a version. Re #12509.

If release name is defined, return this as the version name e.g. if 'releaseName' is 'Windows 10 1511' for build 10586, return 'Windows 10 1511' instead of using release names to builds map and/or Windows Registry. Note that the implied 'else' block is not indented.

josephsl committed:
WinVersion constants: define releae names for all constants. Re #12509.

Although release name texts are useful for Windows 10, apply this change to all constants (Windows 7 and 8.x also).

josephsl committed:
WinVersion: clarify comment. Re #12509.

Note that on server systems, unless otherwise noted, client relesae names are returned. The first such exception is likely the erver counterpart of Windows 10 21H2 (Server 2022) which may have a different build than client releases.

josephsl committed:
WinVersion: actual name for 'releaseName' flag. Re #12509

josephsl committed:
winVersion: cache Windows 10 versions. Re #12509.

Review by Reef Turner (NV Access): in winVersion.getWinVer() function, instead of opening Windows Registry to obtain Windows 10 each time the function is called, have the version string handy from the beginning. Note that Release Id and display version are not found on Windows 10 1507 (build 1024) or earlier.

josephsl commented:
Hi,

For Windows Server 2022, at least "Windows 10 21H2" will be returned, which is fine. The overall PR could change depending on what we learn on June 24th.

Thanks.

josephsl commented:
Hi,

As for unit test, str.startswith should be checked as repr records release name, build, and product type in one go.

Thanks.

josephsl committed:
winVersion/unit tests: add release name test. Re #12509.

Advised by Reef Turner (NV Access): define a unit test to compare release names, specifically to simulate winVersion.getWinVer but with a constant. Because repr(winVersion.getWinVer()) includes release name, major.minor.build, and product type in one go, use asertIn to test the expected string (in this case, 'Windows 10 1607' for Server 2016).

feerrenrut commented on source/winVersion.py:
```diff
@@ -26,35 +27,38 @@ def __init__(
+			# Always return "Windows 10 1507" on build 10240.
+			if self.build == 10240:
+				return "Windows 10 1507"
+			return f"Windows 10 {WIN10CURRENTRELEASENAME}"
```
I don't think this should fall back to the current release name, that would be quite confusing.

feerrenrut replied:
If we want to support manual construction of these classes rather than only using the pre-constructed ones `WIN10_1607` etc, and automatically supplying the release name where it is known, then instead of passing in the name on construction make a build to release name map and look up the release name from the build.

feerrenrut commented on source/winVersion.py:
```diff
@@ -81,22 +85,38 @@ def __ge__(self, other):
+WIN10_21H1 = WinVersion(major=10, minor=0, build=19043, releaseName="Windows 10 21H1")
+
+
+# On Windows 10 1511 and later, cache the version in use on the system.
+with winreg.OpenKey(
+	winreg.HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows NT\CurrentVersion"
+) as currentVersion:
+	# Version 20H2 and later where a separate display version string is used.
+	# Release Id (older releases) and display version (20H2 and later) will be cached.
+	try:
+		WIN10CURRENTRELEASENAME = winreg.QueryValueEx(currentVersion, "DisplayVersion")[0]
+	except OSError:
+		# Don't set anything if this is Windows 10 1507 or earlier.
+		try:
+			WIN10CURRENTRELEASENAME = winreg.QueryValueEx(currentVersion, "ReleaseID")[0]
+		except OSError:
+			WIN10CURRENTRELEASENAME = None
```
Please put this in a function I'd recommend calling it directly from `getWinVer` to supply the `releaseName=` argument when constructing `WinVersion`

josephsl committed:
Revert "winVersion/Windows 10 releases: remove unused releases to builds map. Re #12509."

This reverts commit e5f39aa0ba08a1ed05f2bde7ec2ec92373f743ba.

josephsl committed:
winVersion/releases to builds map: marked for deprecation, to be removed in 2022.1. Re #12509.

Advised by NV Access: mark no longer used releases to builds map as deprecated and to be removed in 2022.1 as Windows 10 release names are obtained from Windows Registry.

josephsl committed:
winVersion: flip release names to builds map with a brand new 'builds to release names' map. Re #12509.

Note from Reef Turner (NV access): flip the now deprecated 'release names to builds' map by defining a new map that records builds to releases. This prepares NVDA to support Windows releases with multiple builds under the same release name, as well as defining the release name for build 10240 (Windows 10 1507) through this mapping instead of hard-coding it in functions.

josephsl committed:
winVersion.WinVersion: transfer bulk of Windows release name fetcher from WinVersion class to a new module-level function. Re #12509.

Addressing a comment from NV Access: use a dedicated function to return Windows release names given its major, minor, and build components. Therefore transfer the bulk of WinVersion classes's Windows release name fetcher to the module and call it from both WinVersion class and getWinVer() function. The string returned will be used as 'releaseName' component of WinVersion class constructor.

josephsl committed:
winVersion.WinVersion class: transfer private Windows release names method to repr method. Re #12509.

Transfer the private Windows release names method in WinVersion class to the repr method as a dedicated module-level function will return Windows release names. In case runtime error is raised, use the name 'Windows release unknown' to make it consistent with winVersion.getWinVer function.

josephsl committed:
winVersion/unit tests: additional tests for Windows release name retrieval functionality. re #12509.

Add the following tests:
* Obtain Windows release name from known Windows release constants defined in winVersion module
* Make sure a runtime error is raised when the new Windows release name function defined in winVersion module receives unknown major.minor/build values
* Windows release name as defined in Windows Registry is returned on any Windows 10 releases not defined in releases to builds map, simulating Windows Insider Preview builds
Also, rename Windows release name test to reflect what it actually tests.

feerrenrut commented on source/winVersion.py:
```diff
@@ -81,32 +70,94 @@ def __ge__(self, other):
+		# Look these up first before asking Windows Registry.
+		if build in BUILDS_TO_RELEASE_NAMES:
+			return BUILDS_TO_RELEASE_NAMES[build]
+		# On Windows 10 1511 and later, cache the version in use on the system.
+		with winreg.OpenKey(
+			winreg.HKEY_LOCAL_MACHINE, r"Software\Microsoft\Windows NT\CurrentVersion"
+		) as currentVersion:
+			# Version 20H2 and later where a separate display version string is used.
+			try:
+				releaseId = winreg.QueryValueEx(currentVersion, "DisplayVersion")[0]
+			except OSError:
+				# Don't set anything if this is Windows 10 1507 or earlier.
+				try:
+					releaseId = winreg.QueryValueEx(currentVersion, "ReleaseID")[0]
+				except OSError:
+					releaseId = "unknown"
+		return f"Windows 10 {releaseId}"
```
Please put this code in a separate function EG `_getRunningVersionNameFromReg`, I don't think it should be called here. It's too easy to conflate the version name for the build given by major, minor, build args with the version name for the currently running version of windows.

In this new function `_getRunningVersionNameFromReg`:
- protect against trying to query the registry before build 1511, raise an exception.
- Cache the result, it shouldn't have to open the registry twice, the running version of windows is not going to change. I'd use [`@lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache).

josephsl replied:
Done. Note that maxsize argument is supplied becasue we are using Python 3.7; in 3.9 there is no need to supply maxsize when defining the decorator.

josephsl commented:
Hi, ah, similar to what Sean is doing with control types refactor? Sure – give me a sec. Thanks for this tip.

josephsl committed:
winVersion/Windows 10 releases to builds (deprecated): officially marked for deprecation (will not be present in the module starting NVDA 2022.1). Re #12509.

Advised by Reef Turner (Nv Access): do not add the deprecated Windows 10 releases to builds map in NVDA 2022.1 or later.

feerrenrut commented on source/winVersion.py:
```diff
@@ -81,32 +70,94 @@ def __ge__(self, other):
+# Records a mapping between Windows builds and release names.
+# These include build 10240 for Windows 10 1507 and releases with multiple release builds.
+# These are applicable to Windows 10 as they report the same system version (10.0).
+BUILDS_TO_RELEASE_NAMES = {
```
Please make this internal 
```suggestion
_BUILDS_TO_RELEASE_NAMES = {
```

feerrenrut commented on source/winVersion.py:
```diff
@@ -81,32 +70,94 @@ def __ge__(self, other):
+WIN8 = WinVersion(major=6, minor=2, build=9200, releaseName="Windows 8")
+WIN81 = WinVersion(major=6, minor=3, build=9600, releaseName="Windows 8.1")
+WIN10 = WIN10_1507 = WinVersion(major=10, minor=0, build=10240, releaseName="Windows 10 1507")
+WIN10_1511 = WinVersion(major=10, minor=0, build=10586, releaseName="Windows 10 1511")
```
Now that we have `BUILDS_TO_RELEASE_NAMES` you can remove this. I'd make it an optional argument on `class WinVersion` which should look it up using the function `_getWindowsReleaseName` but don't fall back to `_getRunningVersionNameFromReg`

feerrenrut commented on source/winVersion.py:
```diff
@@ -81,32 +70,94 @@ def __ge__(self, other):
 	"""
 	winVer = sys.getwindowsversion()
+	try:
+		releaseName = getWindowsReleaseName(major=winVer.major, minor=winVer.minor, build=winVer.build)
```
Try `_getRunningVersionNameFromReg` catch exception, then construct without a release name explicitly set, allowing `class WinVersion` to populate it via `_getWindowsReleaseName`.

josephsl replied:
Yep - the new WinReg method will return an empty string for release Id, which is then processed inside the private release name method. This is actually a stepping stone toward #12585 once this PR is merged into master. Thanks.

josephsl commented on source/winVersion.py:
```diff
@@ -81,32 +70,94 @@ def __ge__(self, other):
+}
+
+
+def getWindowsReleaseName(major: int = 0, minor: int = 0, build: int = 0) -> str:
```
Done - it's a matter of restoring and tweaking the private method from previous commits.

josephsl committed:
winVersion/release name: transfer release name fetcher back to WinVersion class and make it private again. Re #12544.

Noted by NV Access: it is the job of the WinVersion class to inform others about what the release name for a given Windows version should be. Therefore transfer release name fetcher back to WinVersion class and make it private. Also, instead of raising a runtime error, say 'Windows release unknown' if the major.minor.build combination doesn't match anything.

josephsl committed:
winVersion: set release name attribute via private release name method. Re #12544.

Comment from Reef Turner (NV Access): set release name attribute based on info returned from the private Windows version to release name method. Also repr() method was edited to use release name attribute directly, and winVersion.getWinVer() no longer sets release name manually.

josephsl committed:
winVersion: remove release name texts from Windows release constants. Re #12544.

As release name is obtained from Windows and/or predefined, there is no need to store release names for Windows releases.

josephsl committed:
winVersion: privatize and move builds to release names mapping. Re #12544.

Advised by Reef Turner (NV Access): mke builds to release names mapping private, and move this mapping to the top of the module to satisfy Python when running unit tests (Python raises 'name error').

josephsl committed:
winVersion/unit tests: rework release name tests. Re #12544.

Instead of raising a runtime error, the string 'Windows release unknown' is returned if release name could not be determined. Therefore take advantage of this in unit tests.

josephsl committed:
winVersion: split Windows Registry access from release name method. Re #12544.

Split Windows Registry access routine from Windows version to release name method for ease of maintenance. Also, if Registry does not record version name, just return an empty string so the actual release name method can assign something.

josephsl committed:
winVersion/Windows Registry: move the function to the module, protect against older Windows releases without release names defined in Windows Registry. Re #12544.

Move Windows Registry access method from WinVersion class to the module, and:
* Decorate the function to return a cached release name through LRU (least recently used) cache.
* Protect against Windows releases without release names defined in Windows Registry (raises runtime error).

josephsl committed:
Merge branch 'master' of https://github.com/nvaccess/nvda into win10VersionFromWinReg

josephsl committed:
WinVersion.release name: clarify comment. Re #12544

feerrenrut commented on source/winVersion.py:
```diff
@@ -34,21 +78,43 @@ def __init__(
+			# However there are builds with no release name (Version 1507/10240)
+			# or releases with different builds.
+			# Look these up first before asking Windows Registry.
+			if self.build in _BUILDS_TO_RELEASE_NAMES:
+				return _BUILDS_TO_RELEASE_NAMES[self.build]
+			releaseId = _getRunningVersionNameFromWinReg()
+			if not releaseId:
+				releaseId = "unknown"
```
`_getWindowsReleaseName` shouldn't fall back to the registry, the only case where the registry should be used is when getting the version for the version currently running.

```suggestion
			if self.build in _BUILDS_TO_RELEASE_NAMES:
				return _BUILDS_TO_RELEASE_NAMES[self.build]
			if not releaseId:
				releaseId = "unknown"

```

feerrenrut commented on source/winVersion.py:
```diff
@@ -12,6 +12,50 @@
+}
+
+
+@functools.lru_cache(maxsize=128)
```
I think this could only every have one value:
```suggestion
@functools.lru_cache(maxsize=1)

```

feerrenrut commented on source/winVersion.py:
```diff
@@ -70,94 +136,32 @@ def __ge__(self, other):
 	return WinVersion(
 		major=winVer.major,
 		minor=winVer.minor,
 		build=winVer.build,
```
Since this getting the currently running version of windows:
```suggestion
	try:
		releaseName = f"Windows 10 {_getRunningVersionNameFromWinReg()}"
	except RuntimeError:
		releaseName = None
	return WinVersion(
		major=winVer.major,
		minor=winVer.minor,
		build=winVer.build,
		releaseName=releaseName,
```

josephsl replied:
When preparing for #12585, we might as well enhance release name or the private WinReg function to say "Windows 11" if running specific builds.

josephsl commented:
Hi,

Ah, these clarify things a lot - in the future I should sleep before working on PR's...

Thanks.

josephsl committed:
winVersion.WinReg: consult Windows Registry only if retrieivng current version. Re #12544

Co-authored with Reef Turner (NV Access): ask Windows Registry if and only if fetching current reelase.

Co-authored-by: Reef Turner <feerrenrut@users.noreply.github.com>

josephsl committed:
winVersion.WinReg: only cache one value for Relase Id coming from Windows Registry. Re #12509.

Suggested by Reef Turner (NV aCcess): reduce LRU cache overhead by specifying that Registry value should be cached only once.

Co-authored-by: Reef Turner <feerrenrut@users.noreply.github.com>

josephsl committed:
winVErsion.getWinVer: retrieve release name reported by Windows Registry when obtaining current build only. Re #12509.

Suggested by Reef Turner (NV Access): ask Windows Registry when fetching current Windows version only.

josephsl commented:
Hi,

Regarding first suggestion (release Id inside get release name method), since release Id is not used, NVDA will simply say, "Windows 10 unknown". This is done to resolve Flake8 errors, remove unused variable, and to prepare for Windows 11 and future releases.

Thanks.

josephsl committed:
winVersion.releaseName: return 'Windows 10 unknown' if build cannot be determined. re #12509.

On Windows 10 (and later), the following pths will be used to retrieve relase name (in this order):
1. Existing build: whatever the release name is.
2. Current version: Windows Registry. This also covers Windows Insider Preview builds that will eventually become stable (retail) builds.
If none of these work, just say 'Windows 10 unknown'. This also resolves Flake8 F821 (undefined variable name).

josephsl committed:
winVersion: lint, make releaseName parameter optional. Re #12509.

josephsl committed:
Merge remote-tracking branch 'origin/master' into win10VersionFromWinReg

josephsl committed:
winVersion/unit tests: actually check for 'Windows 10 unknown' if testing on public uilds. Re #12509.

Windows Insider Preview builds will say 'Windows 10 Dev', but as unit tests are tested in public builds (unless someone tests in sindier builds), test to see if 'unknown' is returned. Checkig this string is fine because build 21390 is an Insider Preview build that were succeeded by newer builds.

josephsl commented:
Hi, you’re right – thanks for catching this.

josephsl committed:
winVersion/WinReg: typo

josephsl committed:
Merge branch 'master' into win10VersionFromWinReg

lukaszgo1 commented on source/winVersion.py:
```diff
@@ -26,20 +72,26 @@ def __init__(
 			major: int = 0,
 			minor: int = 0,
 			build: int = 0,
+			releaseName: Optional[str] = "",
```
This  is quite unusual. The standard usage for `Optional` is for arguments which are set to `None` by default yet can accept a given type. Here however you're setting it into an empty string as a default value and passing `None` when constructing the instance. Can this be simplified to either always using empty  string and not accepting `None` or preferably using `None` as default value for this kwarg?

lukaszgo1 commented on source/winVersion.py:
```diff
@@ -124,20 +183,23 @@ def isUwpOcrAvailable():
+# Deprecated: Windows 10 releases will be obtained from Windows Registry, no entries will be added.
+# The below map will be removed in 2022.1.
+if version_year < 2022:
+	WIN10_RELEASE_NAME_TO_BUILDS = {
```
Please add a change log entry to the PR description explaining that this map is deprecated.

josephsl commented:
Hi, not really as this map is only used to look up Windows 10 releases, and no-one uses it. I feel it should have been a private map to begin with, but…

josephsl committed:
Merge branch 'master' into win10VersionFromWinReg

josephsl committed:
winVersion.WinVersion/release name: make it 'None' by default. Re #12509

feerrenrut reviewed (changes requested):
Code looks good now.

Let's add support for Windows 11 in a follow up PR.
Can you please update the PR description, including deprecations in the section for the changes file.

josephsl commented:
Hi,

PR intro comment updated with what's new entries (any other changes needed?).

Thanks.

feerrenrut commented:
Yes, looks good @josephsl.

feerrenrut committed:
update changes file for PR #12544

feerrenrut reviewed (approved):
I think it is more useful if the deprecation entry in the change log points to the PR rather than the issue.

josephsl commented:
Hi, agreed. Thanks.
